freqtrade/freqai/RL/BaseEnvironment.py

```diff
@@ -0,0 +1,267 @@
+        self._total_unrealized_profit = 1
+
+        return self._get_observation()
+
+    @abstractmethod
+    def step(self, action: int):
+        """
+        Step depeneds on action types, this must be inherited.
+        """
+        return
+
+    def _get_observation(self):
+        """
+        This may or may not be independent of action types, user can inherit
+        this in their custom "MyRLEnv"
+        """
+        features_window = self.signal_features[(
+            self._current_tick - self.window_size):self._current_tick]
+        features_and_state = DataFrame(np.zeros((len(features_window), 3)),
+                                       columns=['current_profit_pct', 'position', 'trade_duration'],
+                                       index=features_window.index)
+
+        features_and_state['current_profit_pct'] = self.get_unrealized_profit()
+        features_and_state['position'] = self._position.value
+        features_and_state['trade_duration'] = self.get_trade_duration()
+        features_and_state = pd.concat([features_window, features_and_state], axis=1)
+        return features_and_state
+
+    def get_trade_duration(self):
+        """
+        Get the trade duration if the agent is in a trade
+        """
+        if self._last_trade_tick is None:
+            return 0
+        else:
+            return self._current_tick - self._last_trade_tick
+
+    def get_unrealized_profit(self):
+        """
+        Get the unrealized profit if the agent is in a trade
+        """
+        if self._last_trade_tick is None:
+            return 0.
+
+        if self._position == Positions.Neutral:
+            return 0.
+        elif self._position == Positions.Short:
+            current_price = self.add_entry_fee(self.prices.iloc[self._current_tick].open)
+            last_trade_price = self.add_exit_fee(self.prices.iloc[self._last_trade_tick].open)
+            return (last_trade_price - current_price) / last_trade_price
+        elif self._position == Positions.Long:
+            current_price = self.add_exit_fee(self.prices.iloc[self._current_tick].open)
+            last_trade_price = self.add_entry_fee(self.prices.iloc[self._last_trade_tick].open)
+            return (current_price - last_trade_price) / last_trade_price
+        else:
+            return 0.
+
+    @abstractmethod
+    def is_tradesignal(self, action: int):
+        """
+        Determine if the signal is a trade signal. This is
+        unique to the actions in the environment, and therefore must be
+        inherited.
+        """
+        return
+
+    def _is_valid(self, action: int):
+        """
+        Determine if the signal is valid.This is
+        unique to the actions in the environment, and therefore must be
+        inherited.
+        """
+        return
+
+    def add_entry_fee(self, price):
+        return price * (1 + self.fee)
+
+    def add_exit_fee(self, price):
+        return price / (1 + self.fee)
+
```

Comment:
I don't think this is correct.
you're moving price (up or down) by "fee" (same way we do it for regular trading) - but the actual direction will be different if it's a long or short.

Reply:
It is a quite hard question, cuz the fee is generalized here, this "fee" is only for the learning process, and if we would use it directly from the exchange it would get much more complicated, and the possible advantage are very  small. These are nuacnes for the agent, it only counts in the training process, it is just gives the feelings for the agent that there is a fee which will be modify the profits. So in the end, it is easy to use as a generalized fee. 


The answer for the long/short question, it is different but it is different when we actually use this function!

Reply:
long/short is for sure wrong - now if it ain't used - it shouldn't be there (especially since it's not correct).

Reply:
It is correct.

The syntax is terrible, though.

We have an `add_entry_fee` and an `add_exit_fee`. For shorts, we `add_entry_fee` on entrance. For longs we `add_exit_fee` on entrance...yes, the names should be reversed (the calcs are still correct tho). 

I reversed the names so that shorts `add_exit_fee` when they are initiated. Longs `add_entry_fee` when they are initiated. Makes more sense this way.